### PR TITLE
london-10/iryna-lypnyk/completed-mandatory&extra-exercises

### PR DESCRIPTION
**Pull request by IrynaLypnyk**

### CODE REVIEW

File: mandatory/1-syntax-errors.js [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 
-function addNumbers(a b c) {
+function addNumbers(a, b, c) {
   return a + b + c;
 }
 
-function introduceMe(name, age)
-  return `Hello, my {name}` is "and I am $age years old`;
+function introduceMe(name, age) {
+  return `Hello, my name is ${name} and I am ${age} years old`;
```
Comment (Schboostie) [inline mandatory/1-syntax-errors.js:8]: Good use of back tick when adding variables to a string.

File: mandatory/3-function-output.js [hunk 2]
```diff
@@ -1,16 +1,20 @@
 // Add comments to explain what this function does. You're meant to use Google!
+// Returns a random integer from 0 to 9
```
Comment (Schboostie) [inline mandatory/3-function-output.js:2]: You could add further comments to explain how Math.random() returns a decimal value which is then scaled to a desired range.

File: mandatory/4-tax.js [hunk 3]
```diff
@@ -5,7 +5,9 @@
   Sales tax is 20% of the price of the product.
 */
 
-function calculateSalesTax() {}
+function calculateSalesTax(price) {
+  return price + price/5;
```
Comment (Schboostie) [inline mandatory/4-tax.js:9]: It can be useful to declare separate variables at stages of each operation. When calculations become more complexed it may be necessary to check various stages to see if and when mistakes occur.

File: mandatory/4-tax.js [hunk 4]
```diff
@@ -17,7 +19,10 @@ function calculateSalesTax() {}
   Remember that the prices must include the sales tax (hint: you already wrote a function for this!)
 */
 
-function addTaxAndFormatCurrency() {}
+function addTaxAndFormatCurrency(number) {
+  let tax = calculateSalesTax(number);
```
Comment (Schboostie) [inline mandatory/4-tax.js:23]: I really like the use of the tax variable to break down the stages of the function.

File: extra/2-piping.js [hunk 5]
```diff
@@ -16,26 +16,27 @@
...
 
+function format(num) {
+  return `£${num}`
 }
 
 const startingValue = 2;
 
 // Why can this code be seen as bad practice? Comment your answer.
-let badCode =
+let badCode = `£${(startingValue + 10 ) * 2}`
```
Comment (Schboostie) [inline extra/2-piping.js:35]: I can not see a comment to explain why this is bad code. Did you add one?

File: extra/2-piping.js [hunk 6]
```diff
@@ -16,26 +16,27 @@
...
```
Comment (Schboostie) [inline extra/2-piping.js:36]: My explanation was:

 //Multiple methods are used in one line which makes it hard to determine the results from each methods execution. It is confusing and hard to follow.

**Review by Schboostie — COMMENTED** [review]
Hi Iryna, I really liked the way you tackled alot of the problems. I hope you found my comments useful🌟

**Comment by kjjwhitlock** [comment]
Iryna, well done this is a fantastic effort! Really impressed with your work on the extra exercises too. Keep it up!

File: extra/3-magic-8-ball.js [hunk 7]
```diff
@@ -58,10 +61,61 @@ function shakeBall() {
...
+  "My reply is no.",
+  "My sources say no.",
+  "Outlook not so good.",
+  "Very doubtful.",
+];
+
+let allAnswers = [veryPositiveAnswers, positiveAnswers, negativeAnswers, veryNegativeAnswers];
+
+function findRandomIndex(max){
+  return  Math.floor(Math.random() * max);
```
Comment (kjjwhitlock) [inline extra/3-magic-8-ball.js:99]: Great use of max here!

File: extra/3-magic-8-ball.js [hunk 8]
```diff
@@ -58,10 +61,61 @@ function shakeBall() {
...
 }
 
+function shakeBall() {
+  console.log("The ball has shaken!");
+  let num1 = findRandomIndex(4);
+  let num2 = findRandomIndex(5);
+  return allAnswers[num1][num2];
```
Comment (kjjwhitlock) [inline extra/3-magic-8-ball.js:106]: Why are you returning two indices here?

Comment (kjjwhitlock) [reply extra/3-magic-8-ball.js:106]: Ah, I think I see. Do you need to call findRandomIndex within the shakeBall function so that it changes the output each time?

File: extra/3-magic-8-ball.js [hunk 9]
```diff
@@ -58,10 +61,61 @@ function shakeBall() {
...
+ 
+  }
+
+function checkAnswer(answer) {
```
Comment (kjjwhitlock) [inline extra/3-magic-8-ball.js:110]: I am not sure what this is doing (am new to javascript). Could you please explain? Remember to focus on readability. Looks great though and it's awesome to see all the tests passing.

**Review by kjjwhitlock — APPROVED** [review]

File: mandatory/3-function-output.js (hunk 2 above)
Comment (IrynaLypnyk) [reply mandatory/3-function-output.js:2]: I was concentrated so much on JavaScript that I didn`t notice the message about "add comment..." ))) Thank you, Shahid and Katie!

Comment (IrynaLypnyk) [reply mandatory/3-function-output.js:2]: I was concentrated so much on JavaScript that I didn`t notice the message about "add comment..." ))) Thank you, Shahid and Katie!

File: extra/2-piping.js (hunk 5 above)
Comment (IrynaLypnyk) [reply extra/2-piping.js:35]: I concentrated so much on JavaScript that I didn`t notice task "Comment your answer..." ))) I
